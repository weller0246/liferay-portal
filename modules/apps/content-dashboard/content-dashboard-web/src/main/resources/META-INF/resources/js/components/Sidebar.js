/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayButtonWithIcon} from '@clayui/button';
import ClayLayout from '@clayui/layout';
import {useTimeout} from '@liferay/frontend-js-react-web';
import classNames from 'classnames';
import React, {useContext, useEffect, useState} from 'react';

const SidebarContext = React.createContext();

const noop = () => {};

const SidebarBody = ({children, className}) => {
	return (
		<div className={classNames('sidebar-body', className)}>{children}</div>
	);
};

const SidebarHeader = ({title}) => {
	const {onClose} = useContext(SidebarContext);

	return (
		<section className="sidebar-header">
			<ClayLayout.ContentRow className="sidebar-section">
				<ClayLayout.ContentCol expand>
					<h2 className="font-weight-bold mb-0 pr-2 small">
						{title}
					</h2>
				</ClayLayout.ContentCol>

				<ClayLayout.ContentCol>
					<ClayButtonWithIcon
						aria-label="Close"
						className="mt-n2 text-secondary"
						displayType="unstyled"
						onClick={onClose}
						symbol="times"
					/>
				</ClayLayout.ContentCol>
			</ClayLayout.ContentRow>
		</section>
	);
};

const Sidebar = ({children, onClose = noop, open = true}) => {
	const [isOpen, setIsOpen] = useState(false);

	const delay = useTimeout();

	// Wait until the component is rendered to show it so animation happens

	useEffect(() => {
		if (open !== false) {
			delay(() => setIsOpen(true), 100);
		}
		else {
			setIsOpen(false);
		}
	}, [delay, open]);

	useEffect(() => {
		if (isOpen) {
			document.body.classList.add('sidebar-open');
		}
		else {
			document.body.classList.remove('sidebar-open');
		}
	}, [isOpen]);

	return (
		<div className="content-dashboard sidebar sidebar-light sidebar-sm">
			<SidebarContext.Provider value={{onClose}}>
				{children}
			</SidebarContext.Provider>
		</div>
	);
};

Sidebar.Body = SidebarBody;
Sidebar.Header = SidebarHeader;

export {SidebarContext};
export default Sidebar;
