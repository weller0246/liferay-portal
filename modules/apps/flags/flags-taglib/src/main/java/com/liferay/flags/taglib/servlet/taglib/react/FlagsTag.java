/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.flags.taglib.servlet.taglib.react;

import com.liferay.flags.taglib.internal.servlet.ServletContextUtil;
import com.liferay.flags.taglib.internal.servlet.taglib.util.FlagsTagUtil;
import com.liferay.portal.kernel.exception.PortalException;
import com.liferay.portal.kernel.json.JSONObject;
import com.liferay.portal.kernel.json.JSONUtil;
import com.liferay.portal.kernel.language.LanguageUtil;
import com.liferay.portal.kernel.log.Log;
import com.liferay.portal.kernel.log.LogFactoryUtil;
import com.liferay.portal.kernel.model.Company;
import com.liferay.portal.kernel.model.User;
import com.liferay.portal.kernel.theme.ThemeDisplay;
import com.liferay.portal.kernel.util.AggregateResourceBundle;
import com.liferay.portal.kernel.util.HashMapBuilder;
import com.liferay.portal.kernel.util.PortalUtil;
import com.liferay.portal.kernel.util.PortletKeys;
import com.liferay.portal.kernel.util.Validator;
import com.liferay.portal.kernel.util.WebKeys;
import com.liferay.portal.language.LanguageResources;
import com.liferay.taglib.util.IncludeTag;
import com.liferay.taglib.util.TagResourceBundleUtil;

import java.util.Map;
import java.util.ResourceBundle;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.jsp.PageContext;

/**
 * @author Ambrín Chaudhary
 * @author Jorge González
 */
public class FlagsTag extends IncludeTag {

	public String getClassName() {
		return _className;
	}

	public long getClassPK() {
		return _classPK;
	}

	public String getContentTitle() {
		return _contentTitle;
	}

	public String getContentURL() {
		return _contentURL;
	}

	public String getElementClasses() {
		return _elementClasses;
	}

	public String getMessage() {
		return _message;
	}

	public long getReportedUserId() {
		return _reportedUserId;
	}

	public boolean isEnabled() {
		return _enabled;
	}

	public boolean isLabel() {
		return _label;
	}

	public void setClassName(String className) {
		_className = className;
	}

	public void setClassPK(long classPK) {
		_classPK = classPK;
	}

	public void setContentTitle(String contentTitle) {
		_contentTitle = contentTitle;
	}

	public void setContentURL(String contentURL) {
		_contentURL = contentURL;
	}

	public void setElementClasses(String elementClasses) {
		_elementClasses = elementClasses;
	}

	public void setEnabled(boolean enabled) {
		_enabled = enabled;
	}

	public void setLabel(boolean label) {
		_label = label;
	}

	public void setMessage(String message) {
		_message = message;
	}

	@Override
	public void setPageContext(PageContext pageContext) {
		super.setPageContext(pageContext);

		setServletContext(ServletContextUtil.getServletContext());
	}

	public void setReportedUserId(long reportedUserId) {
		_reportedUserId = reportedUserId;
	}

	@Override
	protected void cleanUp() {
		super.cleanUp();

		_className = null;
		_classPK = 0;
		_contentTitle = null;
		_contentURL = null;
		_elementClasses = null;
		_enabled = true;
		_label = true;
		_message = null;
		_reportedUserId = 0;
	}

	@Override
	protected String getPage() {
		return _PAGE;
	}

	@Override
	protected void setAttributes(HttpServletRequest httpServletRequest) {
		try {
			String message = _getMessage();

			httpServletRequest.setAttribute(
				"liferay-flags:flags:data", _getData(message));

			httpServletRequest.setAttribute(
				"liferay-flags:flags:elementClasses", _elementClasses);
			httpServletRequest.setAttribute(
				"liferay-flags:flags:message", message);
			httpServletRequest.setAttribute(
				"liferay-flags:flags:onlyIcon", !_label);
		}
		catch (Exception exception) {
			_log.error(exception, exception);
		}
	}

	private Map<String, Object> _getData(String message)
		throws PortalException {

		return HashMapBuilder.<String, Object>put(
			"context",
			HashMapBuilder.<String, Object>put(
				"namespace", PortalUtil.getPortletNamespace(PortletKeys.FLAGS)
			).build()
		).put(
			"props",
			() -> {
				HttpServletRequest httpServletRequest = getRequest();

				ThemeDisplay themeDisplay =
					(ThemeDisplay)httpServletRequest.getAttribute(
						WebKeys.THEME_DISPLAY);

				return HashMapBuilder.<String, Object>put(
					"baseData", _getDataJSONObject(themeDisplay)
				).put(
					"captchaURI", FlagsTagUtil.getCaptchaURI(httpServletRequest)
				).put(
					"companyName",
					() -> {
						Company company = themeDisplay.getCompany();

						return company.getName();
					}
				).put(
					"disabled", !_enabled
				).put(
					"forceLogin", !FlagsTagUtil.isFlagsEnabled(themeDisplay)
				).put(
					"message",
					() -> {
						if (Validator.isNotNull(message)) {
							return message;
						}

						return null;
					}
				).put(
					"onlyIcon", !_label
				).put(
					"pathTermsOfUse",
					PortalUtil.getPathMain() + "/portal/terms_of_use"
				).put(
					"reasons",
					FlagsTagUtil.getReasons(
						themeDisplay.getCompanyId(), httpServletRequest)
				).put(
					"signedIn", themeDisplay.isSignedIn()
				).put(
					"uri", FlagsTagUtil.getURI(httpServletRequest)
				).build();
			}
		).build();
	}

	private JSONObject _getDataJSONObject(ThemeDisplay themeDisplay) {
		String namespace = PortalUtil.getPortletNamespace(PortletKeys.FLAGS);

		String contentURL = _contentURL;

		if (Validator.isNull(contentURL)) {
			contentURL = FlagsTagUtil.getCurrentURL(getRequest());
		}

		JSONObject dataJSONObject = JSONUtil.put(
			namespace + "className", _className
		).put(
			namespace + "classPK", _classPK
		).put(
			namespace + "contentTitle", _contentTitle
		).put(
			namespace + "contentURL", contentURL
		).put(
			namespace + "reportedUserId", _reportedUserId
		);

		if (themeDisplay.isSignedIn()) {
			User user = themeDisplay.getUser();

			dataJSONObject.put(
				namespace + "reporterEmailAddress", user.getEmailAddress());
		}

		return dataJSONObject;
	}

	private String _getMessage() {
		ResourceBundle resourceBundle = new AggregateResourceBundle(
			TagResourceBundleUtil.getResourceBundle(pageContext),
			LanguageResources.getResourceBundle(
				PortalUtil.getLocale(getRequest())));

		if (Validator.isNotNull(_message)) {
			return LanguageUtil.get(resourceBundle, _message);
		}

		return LanguageUtil.get(resourceBundle, "report");
	}

	private static final String _PAGE = "/flags/page.jsp";

	private static final Log _log = LogFactoryUtil.getLog(FlagsTag.class);

	private String _className;
	private long _classPK;
	private String _contentTitle;
	private String _contentURL;
	private String _elementClasses;
	private boolean _enabled = true;
	private boolean _label = true;
	private String _message;
	private long _reportedUserId;

}