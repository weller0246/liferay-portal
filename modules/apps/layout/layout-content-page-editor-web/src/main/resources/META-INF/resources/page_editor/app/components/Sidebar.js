/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayButtonWithIcon, default as ClayButton} from '@clayui/button';
import ClayIcon from '@clayui/icon';
import ClayLoadingIndicator from '@clayui/loading-indicator';
import {
	ReactPortal,
	useIsMounted,
	useStateSafe,
} from '@liferay/frontend-js-react-web';
import classNames from 'classnames';
import React from 'react';

import useLazy from '../../core/hooks/useLazy';
import useLoad from '../../core/hooks/useLoad';
import usePlugins from '../../core/hooks/usePlugins';
import * as Actions from '../actions/index';
import {config} from '../config/index';
import {useSelectItem} from '../contexts/ControlsContext';
import {useDispatch, useSelector} from '../contexts/StoreContext';
import selectAvailablePanels from '../selectors/selectAvailablePanels';
import {useDropClear} from '../utils/drag-and-drop/useDragAndDrop';
import {useId} from '../utils/useId';

const {Suspense, useCallback, useEffect} = React;

/**
 * Failure to preload is a non-critical failure, so we'll use this to swallow
 * rejected promises silently.
 */
const swallow = [(value) => value, (_error) => undefined];

/**
 * Load the first available panel if the selected sidebar panel ID is not found.
 * This may happen because the list of panels is modified depending on the user permissions.
 *
 * @param {string} panelId
 * @param {Array} panels
 * @param {object} sidebarPanels
 */
const getActivePanelData = ({panelId, panels, sidebarPanels}) => {
	let sidebarPanelId = panelId;

	let panel = panels.some((panel) => panel.includes(sidebarPanelId))
		? sidebarPanels[sidebarPanelId]
		: null;

	if (!panel) {
		sidebarPanelId = panels[0][0];
		panel = sidebarPanels[sidebarPanelId];
	}

	return {panel, sidebarPanelId};
};

export default function Sidebar() {
	const dropClearRef = useDropClear();
	const [hasError, setHasError] = useStateSafe(false);
	const {getInstance, register} = usePlugins();
	const dispatch = useDispatch();
	const isMounted = useIsMounted();
	const load = useLoad();
	const selectItem = useSelectItem();
	const sidebarId = useId();
	const store = useSelector((state) => state);

	const panels = useSelector(selectAvailablePanels(config.panels));
	const sidebarOpen = store.sidebar.open;
	const {panel, sidebarPanelId} = getActivePanelData({
		panelId: store.sidebar.panelId,
		panels,
		sidebarPanels: config.sidebarPanels,
	});

	const promise = panel
		? load(sidebarPanelId, panel.pluginEntryPoint)
		: Promise.resolve();

	const app = {
		Actions,
		config,
		dispatch,
		store,
	};

	let registerPanel;

	if (sidebarPanelId && panel) {
		registerPanel = register(sidebarPanelId, promise, {app, panel});
	}

	const togglePlugin = () => {
		if (hasError) {
			setHasError(false);
		}

		if (registerPanel) {
			registerPanel.then((plugin) => {
				if (
					plugin &&
					typeof plugin.activate === 'function' &&
					isMounted()
				) {
					plugin.activate();
				}
				else if (!plugin) {
					setHasError(true);
				}
			});
		}
	};

	useEffect(
		() => {
			if (panel) {
				togglePlugin(panel);
			}
			else if (sidebarPanelId) {
				dispatch(
					Actions.switchSidebarPanel({
						sidebarOpen: false,
						sidebarPanelId: null,
					})
				);
			}
		},
		/* eslint-disable react-hooks/exhaustive-deps */
		[panel, sidebarOpen, sidebarPanelId]
	);

	useEffect(() => {
		const sideNavigation = Liferay.SideNavigation.instance(
			document.querySelector('.product-menu-toggle')
		);

		if (sideNavigation) {
			const onHandleSidebar = (open) => {
				dispatch(
					Actions.switchSidebarPanel({
						sidebarOpen: open,
					})
				);
			};

			if (!sideNavigation.visible()) {
				onHandleSidebar(true);
			}

			const sideNavigationListener = sideNavigation.on(
				'openStart.lexicon.sidenav',
				() => onHandleSidebar(false)
			);

			return () => {
				sideNavigationListener.removeListener();
			};
		}
	}, []);

	useEffect(() => {
		const wrapper = document.getElementById('wrapper');

		if (!wrapper) {
			return;
		}

		wrapper.classList.add('page-editor__wrapper');
		wrapper.classList.toggle('page-editor__wrapper--padded', sidebarOpen);

		return () => {
			wrapper.classList.remove('page-editor__wrapper');
			wrapper.classList.remove('page-editor__wrapper--padded');
		};
	}, [sidebarOpen]);

	const SidebarPanel = useLazy(
		useCallback(({instance}) => {
			if (typeof instance.renderSidebar === 'function') {
				return instance.renderSidebar();
			}
			else {
				return null;
			}
		}, [])
	);

	const deselectItem = (event) => {
		if (event.target === event.currentTarget) {
			selectItem(null);
		}
	};

	const handleClick = (panel) => {
		const open =
			panel.sidebarPanelId === sidebarPanelId ? !sidebarOpen : true;
		const productMenuToggle = document.querySelector(
			'.product-menu-toggle'
		);

		if (productMenuToggle && !sidebarOpen) {
			Liferay.SideNavigation.hide(productMenuToggle);
		}

		dispatch(
			Actions.switchSidebarPanel({
				sidebarOpen: open,
				sidebarPanelId: panel.sidebarPanelId,
			})
		);
	};

	return (
		<ReactPortal>
			<div
				className="cadmin page-editor__sidebar page-editor__theme-adapter-forms"
				ref={dropClearRef}
			>
				<div
					className={classNames('page-editor__sidebar__buttons', {
						light: true,
					})}
					onClick={deselectItem}
				>
					{panels.reduce((elements, group, groupIndex) => {
						const buttons = group.map((panelId) => {
							const panel = config.sidebarPanels[panelId];

							const active =
								sidebarOpen && sidebarPanelId === panelId;
							const {
								icon,
								isLink,
								label,
								pluginEntryPoint,
								url,
							} = panel;

							if (isLink) {
								return (
									<a
										className={classNames({active})}
										data-tooltip-align="left"
										href={url}
										key={panel.sidebarPanelId}
										title={label}
									>
										<ClayIcon symbol={icon} />
									</a>
								);
							}

							const prefetch = () =>
								load(
									panel.sidebarPanelId,
									pluginEntryPoint
								).then(...swallow);

							return (
								<ClayButtonWithIcon
									aria-pressed={active}
									className={classNames({active})}
									data-tooltip-align="left"
									displayType="unstyled"
									id={`${sidebarId}${panel.sidebarPanelId}`}
									key={panel.sidebarPanelId}
									onClick={() => handleClick(panel)}
									onFocus={prefetch}
									onMouseEnter={prefetch}
									small={true}
									symbol={icon}
									title={label}
								/>
							);
						});

						// Add separator between groups.

						if (groupIndex === panels.length - 1) {
							return elements.concat(buttons);
						}
						else {
							return elements.concat([
								...buttons,
								<hr key={`separator-${groupIndex}`} />,
							]);
						}
					}, [])}
				</div>
				<div
					className={classNames({
						'page-editor__sidebar__content': true,
						'page-editor__sidebar__content--open': sidebarOpen,
						rtl:
							Liferay.Language.direction[
								themeDisplay?.getLanguageId()
							] === 'rtl',
					})}
					onClick={deselectItem}
				>
					{hasError ? (
						<div>
							<ClayButton
								block
								displayType="secondary"
								onClick={() => {
									dispatch(
										Actions.switchSidebarPanel({
											sidebarOpen: false,
											sidebarPanelId:
												panels[0] && panels[0][0],
										})
									);
									setHasError(false);
								}}
								small
							>
								{Liferay.Language.get('refresh')}
							</ClayButton>
						</div>
					) : (
						<ErrorBoundary
							handleError={() => {
								setHasError(true);
							}}
						>
							<Suspense
								fallback={
									<ClayLoadingIndicator
										className="my-4"
										small
									/>
								}
							>
								<SidebarPanel
									getInstance={getInstance}
									pluginId={sidebarPanelId}
								/>
							</Suspense>
						</ErrorBoundary>
					)}
				</div>
			</div>
		</ReactPortal>
	);
}

class ErrorBoundary extends React.Component {
	static getDerivedStateFromError(_error) {
		return {hasError: true};
	}

	constructor(props) {
		super(props);

		this.state = {hasError: false};
	}

	componentDidCatch(error) {
		if (this.props.handleError) {
			this.props.handleError(error);
		}
	}

	render() {
		if (this.state.hasError) {
			return null;
		}
		else {
			return this.props.children;
		}
	}
}
