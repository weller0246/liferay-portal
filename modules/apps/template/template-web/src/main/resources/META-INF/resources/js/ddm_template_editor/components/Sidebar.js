/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {ClayButtonWithIcon} from '@clayui/button';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, {useEffect} from 'react';

import {ElementsSidebarPanel} from './ElementsSidebarPanel';
import PropertiesSidebarPanel from './PropertiesSidebarPanel';
import {PANEL_IDS} from './panelIds';

export default function Sidebar({
	selectedSidebarPanelId,
	setSelectedSidebarPanelId,
}) {
	useEffect(() => {
		const sideNavigation = Liferay.SideNavigation.instance(
			document.querySelector('.product-menu-toggle')
		);

		if (sideNavigation) {
			if (selectedSidebarPanelId && sideNavigation.visible()) {
				sideNavigation.hide();
			}

			const sideNavigationListener = sideNavigation.on(
				'openStart.lexicon.sidenav',
				() => setSelectedSidebarPanelId(null)
			);

			return () => {
				sideNavigationListener.removeListener();
			};
		}
	}, [selectedSidebarPanelId, setSelectedSidebarPanelId]);

	return (
		<div className="ddm_template_editor__App-sidebar">
			<div
				className={classNames(
					'ddm_template_editor__App-sidebar-content',
					{
						open: selectedSidebarPanelId,
					}
				)}
			>
				<ElementsSidebarPanel
					className={classNames({
						'd-none': PANEL_IDS.elements !== selectedSidebarPanelId,
					})}
				/>

				<PropertiesSidebarPanel
					className={classNames({
						'd-none':
							PANEL_IDS.properties !== selectedSidebarPanelId,
					})}
				/>
			</div>
			<div className="ddm_template_editor__App-sidebar-buttons pt-3">
				<ClayButtonWithIcon
					aria-pressed={PANEL_IDS.elements === selectedSidebarPanelId}
					borderless
					className={classNames('mb-2', {
						active: PANEL_IDS.elements === selectedSidebarPanelId,
					})}
					data-tooltip-align="left"
					displayType="secondary"
					monospaced
					onClick={() => {
						setSelectedSidebarPanelId((selectedSidebarPanelId) =>
							selectedSidebarPanelId === PANEL_IDS.elements
								? null
								: PANEL_IDS.elements
						);
					}}
					outline
					symbol="list-ul"
					title={Liferay.Language.get('elements')}
				/>

				<ClayButtonWithIcon
					aria-pressed={
						PANEL_IDS.properties === selectedSidebarPanelId
					}
					borderless
					className={classNames('mb-2', {
						active: PANEL_IDS.properties === selectedSidebarPanelId,
					})}
					data-tooltip-align="left"
					displayType="secondary"
					monospaced
					onClick={() => {
						setSelectedSidebarPanelId((selectedSidebarPanelId) =>
							selectedSidebarPanelId === PANEL_IDS.properties
								? null
								: PANEL_IDS.properties
						);
					}}
					outline
					symbol="cog"
					title={Liferay.Language.get('properties')}
				/>
			</div>
		</div>
	);
}

Sidebar.propTypes = {
	selectedSidebarPanelId: PropTypes.string,
	setSelectedSidebarPanelId: PropTypes.func.isRequired,
};
