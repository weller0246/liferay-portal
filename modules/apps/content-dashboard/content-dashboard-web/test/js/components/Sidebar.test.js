/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {act, cleanup, fireEvent, render} from '@testing-library/react';
import React, {useState} from 'react';

import Sidebar from '../../../src/main/resources/META-INF/resources/js/components/Sidebar';

import '@testing-library/jest-dom/extend-expect';

const ControlledSidebar = ({open, title = ''}) => {
	const [isOpen, setIsOpen] = useState(open);

	return (
		<Sidebar onClose={() => setIsOpen(false)} open={isOpen}>
			<Sidebar.Header title={title} />
		</Sidebar>
	);
};

describe('Sidebar', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.clearAllTimers();
		cleanup();
	});

	it('renders an open sidebar', () => {
		const {container} = render(<Sidebar />);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
		expect(document.body).toHaveAttribute('class', 'sidebar-open');
	});

	it('renders a closed sidebar', () => {
		const {container} = render(<Sidebar open={false} />);

		act(() => {
			jest.runAllTimers();
		});

		expect(container).toMatchSnapshot();
		expect(document.body).not.toHaveAttribute('class', 'sidebar-open');
	});

	it('renders a sidebar with a header with title and subtitle', () => {
		const {container, getByText} = render(
			<Sidebar>
				<Sidebar.Header title="Title" />
			</Sidebar>
		);

		expect(container).toMatchSnapshot();

		expect(getByText('Title')).toBeInTheDocument();
	});

	it('renders a sidebar with body with custom content', () => {
		const {container, getByText} = render(
			<Sidebar>
				<Sidebar.Body>
					<div>Custom content text</div>
				</Sidebar.Body>
			</Sidebar>
		);

		const customChildren = getByText('Custom content text');

		expect(container).toMatchSnapshot();
		expect(customChildren).toBeInTheDocument();
		expect(customChildren.parentNode).toHaveAttribute(
			'class',
			'sidebar-body'
		);
	});

	it('renders an open sidebar with header and close it on close button click', () => {
		const {getByRole} = render(<ControlledSidebar open={true} />);

		act(() => {
			jest.runAllTimers();
		});

		expect(document.body).toHaveAttribute('class', 'sidebar-open');

		const closeIcon = getByRole('presentation');

		expect(closeIcon).toBeInTheDocument();

		fireEvent(
			closeIcon,
			new MouseEvent('click', {
				bubbles: true,
				cancelable: true,
			})
		);

		expect(document.body).not.toHaveAttribute('class', 'sidebar-open');
	});
});
