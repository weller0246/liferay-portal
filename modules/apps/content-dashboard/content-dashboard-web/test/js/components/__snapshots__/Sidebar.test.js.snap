// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders a closed sidebar 1`] = `
<div>
  <div
    class="content-dashboard sidebar sidebar-light sidebar-sm"
  />
</div>
`;

exports[`Sidebar renders a sidebar with a header with title and subtitle 1`] = `
<div>
  <div
    class="content-dashboard sidebar sidebar-light sidebar-sm"
  >
    <section
      class="sidebar-header"
    >
      <div
        class="sidebar-section autofit-row"
      >
        <div
          class="autofit-col autofit-col-expand"
        >
          <h2
            class="font-weight-bold mb-0 pr-2 small"
          >
            Title
          </h2>
        </div>
        <div
          class="autofit-col"
        >
          <button
            aria-label="Close"
            class="mt-n2 text-secondary btn btn-monospaced btn-unstyled"
            type="button"
          >
            <svg
              class="lexicon-icon lexicon-icon-times"
              role="presentation"
            >
              <use
                xlink:href="#times"
              />
            </svg>
          </button>
        </div>
      </div>
    </section>
  </div>
</div>
`;

exports[`Sidebar renders a sidebar with body with custom content 1`] = `
<div>
  <div
    class="content-dashboard sidebar sidebar-light sidebar-sm"
  >
    <div
      class="sidebar-body"
    >
      <div>
        Custom content text
      </div>
    </div>
  </div>
</div>
`;

exports[`Sidebar renders an open sidebar 1`] = `
<div>
  <div
    class="content-dashboard sidebar sidebar-light sidebar-sm"
  />
</div>
`;
